### PR TITLE
bpo-42246: Make sure that line number is correct after a return, as required by PEP 626

### PR DESCRIPTION
**Pull request by markshannon**
This PR ensures that `f_lineno` can be correctly computed after a return, by ensuring that all compiler generated return instructions
have a line number. 

Any block that has no line numbers and has only one predecessor can inherit the line number from its predecessor.
Any block ending in a return has no successors, and thus can be safely duplicated.

By duplicating all blocks that end in a return, but have no line number, those blocks will have only one predecessor and thus can inherit its predecessor's line number, meaning that all implicit returns now have a line number.

For example,
```python
2. if cond:
3.     a
```
currently compiles to:

```
  2           0 LOAD_FAST                0 (cond)
              2 POP_JUMP_IF_FALSE        8

  3           4 LOAD_GLOBAL              0 (a)
              6 POP_TOP
None    >>    8 LOAD_CONST               0 (None)
             10 RETURN_VALUE
```

with this PR it compiles to:

```
  2           0 LOAD_FAST                0 (cond)
              2 POP_JUMP_IF_FALSE       12

  3           4 LOAD_GLOBAL              0 (a)
              6 POP_TOP
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE

  2     >>   12 LOAD_CONST               0 (None)
             14 RETURN_VALUE
```

which is equivalent, but produces the correct line number when `not cond`.

https://bugs.python.org/issue42246

### CODE REVIEW

**Review by gpshead — APPROVED** [review]
LGTM though it'd be a good idea to have more eyeballs than just mine on this. 👍

**Comment by markshannon** [comment]
@pablogsal would you mind taking a quick look at this?

**Comment by pablogsal** [comment]
> @pablogsal would you mind taking a quick look at this?

Yep, will review this today

File: Python/compile.c [hunk 1]
```diff
@@ -6432,9 +6502,63 @@ optimize_cfg(struct assembler *a, PyObject *consts)
...
        }
     }
+    minimize_lineno_table(a);
     return 0;
 }
 
+static int
+is_exit_without_lineno(basicblock *b) {
+    return b->b_exit && b->b_instr[0].i_lineno < 0;
+}
```
Comment (pablogsal) [inline Python/compile.c]:
```suggestion
static inline int
is_exit_without_lineno(basicblock *b) {
    return b->b_exit && b->b_instr[0].i_lineno < 0;
}
```

File: Python/compile.c [hunk 2]
```diff
@@ -812,6 +812,28 @@ compiler_use_next_block(struct compiler *c, basicblock *block)
...
+{
+    /* Cannot copy a block if it has a fallthrough, since
+     * a block can only have one fallthrough predecessor.
+     */
+    assert(block->b_nofallthrough);
+    basicblock *result = compiler_next_block(c);
+    if (result == NULL) {
+        return NULL;
+    }
+    for(int i = 0; i < block->b_iused; i++) {
```
Comment (pablogsal) [inline Python/compile.c]:
```suggestion
    for (int i = 0; i < block->b_iused; i++) {
```

File: Python/compile.c [hunk 3]
```diff
@@ -6358,16 +6398,32 @@ normalize_basic_block(basicblock *bb) {
...
+                    bb->b_instr[i].i_target = bb->b_instr[i].i_target->b_next;
+                }
+
         }
     }
     return 0;
 }
 
 
+static int
```
Comment (pablogsal) [inline Python/compile.c]:
```suggestion
static inline int
```

File: Python/compile.c [hunk 4]
```diff
@@ -2747,6 +2770,7 @@ compiler_for(struct compiler *c, stmt_ty s)
     ADDOP(c, GET_ITER);
     compiler_use_next_block(c, start);
     ADDOP_JUMP(c, FOR_ITER, cleanup);
+    compiler_use_next_block(c, body);
```
Comment (pablogsal) [inline Python/compile.c:2773]: Why is this needed? To repeat the "body" if there is an implicit return? If so, could you add an example to the test?

File: Python/compile.c [hunk 5]
```diff
@@ -6432,9 +6502,63 @@ optimize_cfg(struct assembler *a, PyObject *consts)
...
+ * copy the line number from the sole predecessor block.
+ */
+static int
+ensure_exits_have_lineno(struct compiler *c)
+{
+    /* Copy all exit blocks without line number that are targets of a jump.
+     */
+    for (basicblock *b = c->u->u_blocks; b != NULL; b = b->b_list) {
+        if (b->b_iused > 0 && is_jump(&b->b_instr[b->b_iused-1])) {
+            switch(b->b_instr[b->b_iused-1].i_opcode) {
```
Comment (pablogsal) [inline Python/compile.c]:
```suggestion
            switch (b->b_instr[b->b_iused-1].i_opcode) {
```

**Comment by pablogsal** [comment]
LGTM in general but I left some minor suggestions

File: Python/compile.c (hunk 3 above)
Comment (markshannon) [reply Python/compile.c]: I've deleted the function and inlined its body of the function. It was only used in once place.

File: Python/compile.c (hunk 4 above)
Comment (markshannon) [reply Python/compile.c:2773]: `FOR_ITER` is a branch, so we need a new block immediately after it.
https://github.com/python/cpython/pull/23495/files#diff-ebc983d9f91e5bcf73500e377ac65e85863c4f77fd5b6b6caf4fcdf7c0f0b057R6401 checks it.
